# Práctica Final de NLP

# 4. Reporte de métricas y conclusiones

## 4.a. Modelo de Machine Learning - Logistic Regression

* Cargamos el modelo

In [ ]:
from tensorflow.keras.models import load_model
cache_dir = '/content/drive/My Drive/0. KEEPCODING/NLP/'
model_file = "ML_model.h5"
model = load_model(os.path.join(cache_dir, model_file))

* Evaluamos el modelo

In [53]:
y_test_pred = best_model.predict(x_test)

# Métricas finales
print("Accuracy en test:", accuracy_score(y_test, y_test_pred))
print("\nReporte de clasificación:\n", classification_report(y_test, y_test_pred))
print("\nMatriz de confusión:\n", confusion_matrix(y_test, y_test_pred))

Accuracy en test: 0.86

Reporte de clasificación:
               precision    recall  f1-score   support

           0       0.88      0.84      0.86        75
           1       0.85      0.88      0.86        75

    accuracy                           0.86       150
   macro avg       0.86      0.86      0.86       150
weighted avg       0.86      0.86      0.86       150


Matriz de confusión:
 [[63 12]
 [ 9 66]]


## 4.b. Modelo de Deep Learning - LSTM

* Cargamos el modelo

In [ ]:
from tensorflow.keras.models import load_model
cache_dir = '/content/drive/My Drive/0. KEEPCODING/NLP/'
model_file = "DL_model.h5"
model = load_model(os.path.join(cache_dir, model_file))

* Evaluamos el modelo

In [56]:
scores = model.evaluate(x_test, y_test, verbose=0)
print("Test accuracy:", scores[0])

Test accuracy: 0.6931491494178772


## 4.c. Conclusiones

El objetivo de esta práctica ha sido principalmente practicar la manipulación y transformación de texto, así como su utilización para generar modelos tanto de Machine Learning como de Deep Learning. Pero el objetivo nunca fue obtener un modelo perfectamente funcional con unos resultados sobresalientes.
Esto se debe principalmente a la limitación en procesamiento, obligándonos a trabajar con una cantidad de datos reducida.
Además, la fase de preprocesamiento debería ser mucho más extensa y meticulosa, filtrando mayor cantidad de palabras entre otros.
A pesar de ello, con las herramientas de las que disponemos, se ha tratado de conseguir los mejores resultados posibles.
Por un lado, el modelo de Machine Learning mediante Logistic Regression nos ha ofrecido unos resultados bastante buenos, con una precisión del 0.85 para reviews negativas y 0.88 para reviews positivas, lo cual supone un buen porcentaje de acierto, y en ambos casos con un f1-score de 0.88.
Por otro lado, el modelo de Deep Learning no ha obtenido unos resultados tan notables, aunque sí aceptables dadas las condiciones, con un accuracy cercano a 0.7.
En definitiva, considerando los recursos que disponemos y basándonos en los resultados, sería más conveniente centrarnos en mejorar el modelo de Machine Learning debido a su simplicidad y efectividad. No obstante, con más tiempo y recursos sería muy recomendable seguir trabajando con el modelo de Deep Learning. Esto se debe a que trabajando con mayor precisión la fase de preprocesamiento, y aumentando la capacidad del modelo mediante la adición de capas neuronales y modificación de otros hiperparámetros, considero que se podrían obtener resultados mucho mejores.